Add progressive redaction (#246)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

for signature verification using reasonable and appropriate means for the given
key-distribution method.

> ## Selective Disclosure without Holder Binding

Reviewer: This is a somewhat nondescript section title and feels duplicate to the "Holder Binding" title above; how about this?
```suggestion
## Forwarding Credentials
```

Author: @Sakurann before I accept, what do you think?

Reviewer 2: I think this is relevant to #254. Is `Credentials` a term used for SD-JWT + disclosures..?

The Issuers and the Verifiers should be aware that when Holder Binding is not present,

> any entity that is not the original receiver can be presenting the Combined Format for Presentation.

Author: I am accepting this change it makes sense to me.

> When doing so, that entity may be applying selective disclosure to the SD-JWT
> that has already been previously selectively disclosed.

Reviewer: ```suggestion
When doing so, that entity may remove Disclosures such that the receiver
learns only a subset of the claims contained in the original Combined Format for Presentation.
```

Author: This seems better, maybe position "receiver" in terms of a "next verifier" to emphasis the "enforcement" comment you made earlier.

> ## Progressive Redaction
> 
> Issuers should be aware that when holder binding is not present,
> documents can be progressively redacted by any holder,
> and further redaction can be applied to an already disclosed token.
> 
> For example, a device manufacturer might produce an SD-JWT
> containing information about upstream and downstream supply chain contributors.
> 
> Each supply chain party can verify only the disclosed claims
> they have received from an upstream party, and they can
> choose to redact claims before presenting the redacted token to a downstream party.
> 
> In some scenarios this behavior could be desirable,
> but if it is not, issuers are encouraged to apply holder binding.

Reviewer 2: ```suggestion
## Selective Disclosure without Holder Binding

The Issuers and the Verifiers should be aware that when Holder Binding is not present,
any entity that is not the original receiver can be presenting the Combined Format for Presentation.
When doing so, that entity may be applying selective disclosure to the SD-JWT
that has already been previously selectively disclosed.

For example, a device manufacturer might produce an SD-JWT
containing information about upstream and downstream supply chain contributors.
Each supply chain party can verify only the claims that were selectively disclosed to them
by an upstream party, and they can choose to apply selective disclosure to those claims
when presenting to a downstream party.

In some scenarios this behavior could be desirable,
but if it is not, the Issuers are encouraged to apply Holder Binding.
```
I like the direction of the text, tried to paraphrase so that already defined terms are re-used.
Unless progressive redaction" is a term well-defined in supply chain industry..?

> when presenting to a downstream party.

Reviewer 2: whitespace in the end of line 794 needs to be removed or won't compile

> when presenting to a downstream party.

Reviewer 2: I would prefer `further remove`. further restrict means something different to me.

> when presenting to a downstream party.

Reviewer: "further reduce"?

> by an upstream party, and they can choose to apply selective disclosure to those claims
> when presenting to a downstream party.

Reviewer: ```suggestion
by an upstream party, and they can choose to further reduce the disclosed claims
when presenting to a downstream party.
```

> In some scenarios this behavior could be desirable,
> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer 3: ```suggestion
In some scenarios this behavior could be desirable,
but if it is not, Verifiers are encouraged to enforce Holder Binding.
```
maybe not use a MUST here?

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer: or "need to"?

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer 3: "need to" would work too. I was just wanting to avoid introducing the capitalized normative language in this section that is about general conceptual guidance.

> In some scenarios this behavior could be desirable,
> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer: ```suggestion
In some scenarios this behavior could be desirable,
but if it is not, Verifiers need to enforce Holder Binding.
```

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Author: maybe "Issuers ought to (provide for / leverage / require ), and Verifiers ought to enforce Holder Binding".

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer 2: I think the point is **the Issuers** need to enforce holder binding if it does not want anyone to be able to forward sd-jwts all over. it's not about verifiers ignoring holder binding when it is present.

(I am ok with `need to enforce` but it should be issuers, not verifiers)

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer: How can an Issuer enforce holder binding when the Verifier can always opt not to check it?

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Author: Issuer needs to be aware that without binding, the credential is basically a Netflix password in a big family.
Verifier (Netflix) needs to be aware that accepting a password with no binding maybe means the current user is not the original account holder.

Sorry for trollish analogy, but they both need to be given guidance (Issuer and Verifier).

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Reviewer: Yes, both need guidance, but the Issuer cannot _enforce_ holder binding. Only the Verifier can. The Issuer should _support_ holder binding though.

> but if it is not, the Issuers are encouraged to apply Holder Binding.

Author: The issuer "enables" holder binding", the verifier "enforces" it.
